<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-проекта" data-toc-modified-id="Описание-проекта-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание проекта</a></span><ul class="toc-item"><li><span><a href="#Задачи" data-toc-modified-id="Задачи-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Задачи</a></span></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Описание данных</a></span></li></ul></li><li><span><a href="#Загрузка-и-обзор-данных" data-toc-modified-id="Загрузка-и-обзор-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Загрузка и обзор данных</a></span></li><li><span><a href="#Количество-книг" data-toc-modified-id="Количество-книг-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Количество книг</a></span><ul class="toc-item"><li><span><a href="#Результат" data-toc-modified-id="Результат-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Результат</a></span></li></ul></li><li><span><a href="#Количество-обзоров-и-средняя-оценка" data-toc-modified-id="Количество-обзоров-и-средняя-оценка-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Количество обзоров и средняя оценка</a></span><ul class="toc-item"><li><span><a href="#Результат" data-toc-modified-id="Результат-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Результат</a></span></li></ul></li><li><span><a href="#Лидирующее-издательство" data-toc-modified-id="Лидирующее-издательство-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Лидирующее издательство</a></span><ul class="toc-item"><li><span><a href="#Результат" data-toc-modified-id="Результат-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Результат</a></span></li></ul></li><li><span><a href="#Самый-рейтинговый-автор" data-toc-modified-id="Самый-рейтинговый-автор-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Самый рейтинговый автор</a></span><ul class="toc-item"><li><span><a href="#Результат" data-toc-modified-id="Результат-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Результат</a></span></li></ul></li><li><span><a href="#Среднее-количество-обзоров" data-toc-modified-id="Среднее-количество-обзоров-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Среднее количество обзоров</a></span><ul class="toc-item"><li><span><a href="#Результат" data-toc-modified-id="Результат-7.1"><span class="toc-item-num">7.1&nbsp;&nbsp;</span>Результат</a></span></li></ul></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Выводы</a></span></li></ul></div>

# Анализ базы данных сервиса для чтения книг по подписке

## Описание проекта
Компания купила крупный сервис для чтения книг по подписке. Необходимо проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Данные нужны, чтобы сформулировать ценностное предложение для нового продукта.

### Задачи
* Посчитать, сколько книг вышло после 1 января 2000 года;
* Для каждой книги посчитать количество обзоров и среднюю оценку;
* Определить издательство, которое выпустило наибольшее число книг толще 50 страниц, чтобы исключить из анализа брошюры;
* Определить автора с самой высокой средней оценкой книг. Учитывать только книги с 50 и более оценками;
* Посчитать среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

### Описание данных
Таблица: `books`. Содержит данные о книгах: 
- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

Таблица `authors`. Содержит данные об авторах:
- `author_id` — идентификатор автора;
- `author` — имя автора.

Таблица `publishers`. Содержит данные об издательствах:
- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

Таблица `ratings`. Содержит данные о пользовательских оценках книг:
- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

Таблица `reviews`. Содержит данные о пользовательских обзорах на книги:
- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Загрузка и обзор данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [3]:
# обзор датаесета books
query = '''SELECT * FROM books LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
# обзор датаесета authors
query = '''SELECT * FROM authors LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [5]:
# обзор датаесета publishers
query = '''SELECT * FROM publishers LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [6]:
# обзор датаесета ratings
query = '''SELECT * FROM ratings LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [7]:
# обзор датаесета reviews
query = '''SELECT * FROM reviews LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

## Количество книг
*Задача: посчитать, сколько книг вышло после 1 января 2000 года*

In [8]:
query = '''
SELECT COUNT(book_id)
FROM books
WHERE CAST(publication_date AS date) > '2000-01-01'
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count
0    819

In [9]:
# количество вышедших книг включая 1.01.2000
query = '''
SELECT COUNT(book_id)
FROM books
WHERE CAST(publication_date AS date) > '1999-12-31'
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count
0    821

### Результат
После 1 января 2000 года вышло 819 книг, еще 2 книги были выпущены 1го января.

## Количество обзоров и средняя оценка
*Задача: для каждой книги посчитать количество обзоров и среднюю оценку*

In [10]:
#  количество обзоров и средняя оценка для каждой книги
query = '''
WITH review_count AS (SELECT book_id,
                             COUNT(review_id) AS review_cnt
                      FROM reviews
                      GROUP BY book_id),
     rating_avg AS (SELECT book_id,
                    ROUND(AVG(rating),2) AS rating_avg
                    FROM ratings
                    GROUP BY book_id)                      
SELECT b.title,
       rc.review_cnt,
       rt.rating_avg
FROM books AS b
LEFT OUTER JOIN review_count AS rc ON b.book_id=rc.book_id
LEFT OUTER JOIN rating_avg AS rt ON b.book_id=rt.book_id
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

title  review_cnt  rating_avg
0            The Body in the Library (Miss Marple  #3)         2.0        4.50
1                                            Galápagos         2.0        4.50
2                             A Tree Grows in Brooklyn         5.0        4.25
3    Undaunted Courage: The Pioneering First Missio...         2.0        4.00
4                                          The Prophet         4.0        4.29
..                                                 ...         ...         ...
995   The Cat in the Hat and Other Dr. Seuss Favorites         NaN        5.00
996    Anne Rice's The Vampire Lestat: A Graphic Novel         NaN        3.67
997                          Essential Tales and Poems         NaN        4.00
998                               Leonardo's Notebooks         NaN        4.00
999                            The Natural Way to Draw         NaN        3.00

[1000 rows x 3 columns]

In [11]:
# Средние и крайние значения количества обзоров и средней оценки
query = '''
SELECT ROUND(AVG(a.review_cnt)) AS rev,
       MIN(a.review_cnt) AS min_rev_cnt,
       MAX(a.review_cnt) AS max_rev_cnt,
       MIN(a.rating_avg) AS min_rat,
       MAX(a.rating_avg) AS max_rat
FROM (
    WITH review_count AS (SELECT book_id,
                                 COUNT(review_id) AS review_cnt
                          FROM reviews
                          GROUP BY book_id),
         rating_avg AS (SELECT book_id,
                               ROUND(AVG(rating),2) AS rating_avg
                        FROM ratings
                        GROUP BY book_id)                      
    SELECT b.title,
           rc.review_cnt,
           rt.rating_avg
    FROM books AS b
    LEFT OUTER JOIN review_count AS rc ON b.book_id=rc.book_id
    LEFT OUTER JOIN rating_avg AS rt ON b.book_id=rt.book_id) AS a
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

rev  min_rev_cnt  max_rev_cnt  min_rat  max_rat
0  3.0            1            7      1.5      5.0

### Результат
По итогам рассчета количества обзоров и среднего рейтинга для 999 книг определено, что всреднем на книгу написано 3 обзора (разброс от 1 до 7 штук), средний рейтинг от 1.5 до 5 баллов.

## Лидирующее издательство
*Задача: определить издательство, которое выпустило наибольшее число книг толще 50 страниц, чтобы исключить из анализа брошюры*

In [12]:
query = '''
SELECT p.publisher,
       COUNT(b.book_id)
FROM books AS b
JOIN publishers AS p ON b.publisher_id=p.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY COUNT(b.book_id) DESC
LIMIT 1
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher  count
0  Penguin Books     42

### Результат
Издательство, которое выпустило больше всего книг - Penguin Books, всего было издано 42 книги, при этом брошюры до 50 страниц не учтены.

## Самый рейтинговый автор
*Задача: определить автора с самой высокой средней оценкой книг. Учитывать только книги с 50 и более оценками*

In [13]:
query = '''
SELECT a.author,
       ROUND(AVG(r.rating),2)
FROM books AS b
LEFT OUTER JOIN ratings AS r ON b.book_id=r.book_id
LEFT OUTER JOIN authors AS a ON b.author_id=a.author_id
WHERE b.book_id IN (SELECT book_id 
                    FROM ratings
                    GROUP BY book_id
                    HAVING COUNT(rating_id) >= 50)
GROUP BY a.author, b.author_id 
ORDER BY AVG(r.rating) DESC
LIMIT 1
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author  round
0  J.K. Rowling/Mary GrandPré   4.29

### Результат
Автор с самой высокой средней оценкой книг среди книг оцененных от 50ти раз - Джоан Роулинг.

## Среднее количество обзоров
*Задача: посчитать среднее количество обзоров от пользователей, которые поставили больше 48 оценок*

In [14]:
query = '''
SELECT ROUND(AVG(review_cnt))
FROM (SELECT username,
             COUNT(review_id) AS review_cnt
      FROM reviews
      WHERE username IN (SELECT username
                         FROM ratings
                         GROUP BY username
                         HAVING COUNT(rating_id) > 48)
      GROUP BY username) AS a
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

round
0   24.0

### Результат
Среднее количество обзоров от пользователей, которые поставили больше 48 оценок - 24 штуки.

## Выводы
База данных сервиса для чтения книг по подписке представленна пятью таблицами, отдельно собрана информация о книгах, об авторах, об издательствах, о пользовательских оценках книг и о пользовательских обзорах на книги. В результате анализа рассчитано количество обзоров и средняя оценка на каждую книгу и определено, что:
* после 1 января 2000 года вышло 819 книг;
* среднее количество обзоров на одну книгу - 3;
* рейтинг варьируется от 1.5 до 5 баллов;
* больше всего книг выпустили Penguin Books, 42 книги;
* автор с самой высокой средней оценкой - Джоан Роулинг;
* самые активные пользователи в среднем оставляют 24 обзора.